libraries/AC_AttitudeControl/AC_WeatherVane.cpp

```diff
@@ -0,0 +1,235 @@
+    // @Description: Below this hoizontal velocity weathervaning is permitted.  Based on ground speed.  Set to 0 to ignore this condition when checking if vehicle should weathervane.
+    // @Units: m/s
+    // @Range: 0 50
+    // @Increment: 0.1
+    // @User: Standard
+    AP_GROUPINFO("VXY_MAX", 5, AC_WeatherVane, _max_vel_xy, 2.0),
+
+    // @Param: VZ_MAX
+    // @DisplayName: Weathervaning max vertical speed
+    // @Description: The maximum climb or descent speed that the vehicle will still attempt to weathervane.  Set to 0 to ignore this condition to get the aircraft to weathervane at any climb/descent rate.
+    // @Units: m/s
+    // @Range: 0 5
+    // @Increment: 0.1
+    // @User: Standard
+    AP_GROUPINFO("VZ_MAX", 6, AC_WeatherVane, _max_vel_z, 1.0),
+
+    AP_GROUPEND
+};
+
+
+// Constructor
+AC_WeatherVane::AC_WeatherVane(const AP_InertialNav& inav) :
+    _inav(inav)
+{
+    reset();
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+float AC_WeatherVane::get_yaw_rate_cds(const int16_t roll_cdeg, const int16_t pitch_cdeg)
+{
+    if (!active_msg_sent) {
+        gcs().send_text(MAV_SEVERITY_INFO, "Weathervane Active");
+        active_msg_sent = true;
+    }
+
+    float output = 0.0;
+    switch (get_direction()) {
+        case Direction::OFF:
+            return 0.0;
+
+        case Direction::NOSE_OR_TAIL_IN:
+            output = roll_cdeg * _gain;
+
+            if (pitch_cdeg > 0) {
+                output *= -1.0;
+            }
+            break;
+
+        case Direction::NOSE_IN:
+            output = (fabsf(roll_cdeg) + MAX(pitch_cdeg,0.0)) * _gain;
+
+            // Yaw in the direction of the lowest 'wing'
+            if (roll_cdeg < 0) {
+                output *= -1.0;
+            }
+            break;
+
+        case Direction::SIDE_IN:
+            output = pitch_cdeg * _gain;
+
+            if (roll_cdeg < 0) {
+                output *= -1.0;
+            }
+            break;
+
+    }
+
+    // Don't activate weather vane if less than deadzone angle
+    float angle = get_direction() == Direction::SIDE_IN ? pitch_cdeg : roll_cdeg;
+    if (fabsf(angle) < _min_dz_ang_deg*100.0 && !(pitch_cdeg > _min_dz_ang_deg*100.0 && get_direction() == Direction::NOSE_IN)) {
+        output = 0.0;
+    }
+
+    // Force the controller to relax.  This can be called when landing.
+    if (should_relax) {
+        output = 0.0;
+        // Always reset should relax. Maintain relaxed condition by persistant calls to relax()
+        should_relax = false;
+    }
+
+    // Slew output
+    last_output = 0.98 * last_output + 0.02 * output;
+
+    return last_output;
+}
+
+// Called on an interupt to reset the weathervane controller
+void AC_WeatherVane::reset(void)
+{
+    last_output = 0;
+    active_msg_sent = false;
+    should_relax = false;
+    first_activate_ms = 0;
+}
+
+bool AC_WeatherVane::should_weathervane(const int16_t pilot_yaw, const int16_t roll_cdeg, const int16_t pitch_cdeg)
```

Comment:
the pilot_yaw input is a bit vehicle specific

Reply:
pilot_yaw is to check if we have had a pilot's command to yaw ti make sure we aren't fighting the pilot.  The variable can be an angle, rate, or rcin.  We just need to check it is non-zero.  So the name seems appropriate to me, but I open to other suggestions.